Add Farming Simulator 2022/2019

Vito0912 opened the pull request:
Added Farming Simulator 2022. It may also work with older versions, although I currently only own Farming Simulator 19 and 17. I will need to investigate whether I can set up a server for them. As of now, this has only been tested with a Farming Simulator 2022 server.

Note:
Server queries are not possible. To obtain the necessary token, you must own or have permission to access the web interface.

I have not tested whether the `token` works when used in code; my testing was limited to using it with the command line.

Vito0912 commented:
I am not able to start a Farming Simulator 17 Server. It will likely work, but I cannot test it.

CosminPerRam commented on GAMES_LIST.md:
```diff
@@ -460,6 +461,10 @@ Conan Exiles never responds to player query.
 Many Minecraft servers do not respond with players data.
 
+### <a name='farmingsimulator'>Farming Simulator
+Farming Simulator 2022 servers need a token (reffered as code in the game). It can be obtained at your server's web interface (http://ip:port/settings.html). It can be passed to GameDig with the additional option: `token`
```
`Farming Simulator 2022 servers need a token`, so, only FS22 or both? as this note is added on both of them.

CosminPerRam commented on lib/QueryRunner.js:
```diff
@@ -102,7 +102,7 @@ export default class QueryRunner {
       }
     }
-
+    
```
This line doesnt need to be changed.

CosminPerRam commented on lib/games.js:
```diff
@@ -1186,6 +1186,22 @@ export const games = {
   },
+  farmingsimulator2019: {
+    name: 'Farming Simulator 2019',
+    release_year: 2021,
```
Should be 2018.

CosminPerRam commented:
As the queries require tokens to be made, you think you could join the Discord server or message me on Discord regarding providing me the needed ones? As I dont own the game I cannot really verify that this PR is valid (not that I don't trust you, but its just to make sure all is good).

Vito0912 commented:
> So as the changes in this PR, we would need to also test FS19.

I tested it with Farming Simulator 19

CosminPerRam commented on protocols/farmingsimulator.js:
```diff
@@ -0,0 +1,54 @@
+      }
+    });
+
+    state.mods = [];
```
`mods` is not a stabilized field, so it should be under `state.raw`.

CosminPerRam commented on GAMES_LIST.md:
```diff
@@ -122,6 +122,8 @@
 | factorio             | Factorio                                         |                                                 |
 | farcry               | Far Cry                                          |                                                 |
 | farcry2              | Far Cry 2                                        |                                                 |
+| farmingsimulator2019 | Farming Simulator 2019                           | [Notes](#farmingsimulator)                      |
```
The name of the game should be `Farming Simulator 19`, same for 22.

Vito0912 replied:
Should the naming of the key also be changed?

CosminPerRam replied:
Yep.

CosminPerRam replied:
(Requesting the name change as seeing on their website that they call the game "Farming Simulator 22", not "Farming Simulator 2022")

CosminPerRam replied:
(same goes for the names in the `games.js` file).

CosminPerRam reviewed (changes requested):
Privately validated that the queries work.
Please also run eslint (:

podrivo commented:
@Vito0912 Sorry, but still a conflict to resolve before merging...

CosminPerRam reviewed (approved):
Good to merge, but please also run `eslint` of the protocol file.

podrivo commented:
@Vito0912 Do we have a `version` field in the query response? Wondering if we should add a `version` top level for Farming Simulator as well.

Vito0912 commented:
@podrivo Yeah, I am pretty sure it does have a version in it's response. Can't remember that there was a top level `version`. I will add it as soon as I have time.

podrivo commented:
No sweat!
If you could add a query response at the original post, I'll make sure to add it in the other PR (https://github.com/gamedig/node-gamedig/pull/532).

Vito0912 commented:
I've updated the file to adhere to the ESLint standards and have also added the version information to raw.version. I'm uncertain as to why it triggered another merge of the branch. I hope this won't cause any issues.
